File: web/src/commands/recordings.ts
import type { LocalFileState, RecordingFile } from '../db/recordings'
import type { CrosstuneDb } from '../db/schema'
import type { LocalRecording } from '../db/types'
import { RECORDING_NOT_FOUND, TUNE_NOT_FOUND } from './messages'
import { activeByPosition, newId, nextPosition, now, putRow, recordingTx, tombstone } from './write'

function emptyFile(id: string, overrides: Partial<RecordingFile> = {}): RecordingFile {
  return {
    id,
    blob: null,
    mime: null,
    bytes: 0,
    local_duration_ms: null,
    local_state: 'captured',
    error: null,
    last_chunk_at: null,
    tune_id: null,
    recorded_at: null,
    next_attempt_at: null,
    upload_attempts: 0,
    ...overrides,
  }
}

export function activeRecordingsForTune(
  db: CrosstuneDb,
  tuneId: string,
): Promise<LocalRecording[]> {
  return db.recordings.where('tune_id').equals(tuneId).toArray().then(activeByPosition)
}

async function putRecordingRow(
  db: CrosstuneDb,
  id: string,
  fields: {
    tuneId: string | null
    source: 'microphone' | 'upload'
    label: string | null
    recordedAt: string
  },
): Promise<void> {
  const at = now()
  const siblings = fields.tuneId ? await activeRecordingsForTune(db, fields.tuneId) : []
  await putRow(db, 'recordings', {
    id,
    created_at: at,
    updated_at: at,
    deleted_at: null,
    server_seq: 0,
    tune_id: fields.tuneId,
    label: fields.label,
    source: fields.source,
    recorded_at: fields.recordedAt,
    position: nextPosition(siblings),
    state: 'pending_upload',
    duration_ms: null,
    playback_mime: null,
    playback_bytes: null,
    error: null,
  })
}

export async function beginCapture(
  db: CrosstuneDb,
  id: string,
  fields: { tuneId: string | null; recordedAt: string },
): Promise<void> {
  await db.recording_files.put(
    emptyFile(id, {
      local_state: 'capturing',
      last_chunk_at: Date.now(),
      tune_id: fields.tuneId,
      recorded_at: fields.recordedAt,
    }),
  )
}

export async function appendChunk(
  db: CrosstuneDb,
  id: string,
  idx: number,
  blob: Blob,
): Promise<void> {
  await db.transaction('rw', db.recording_files, db.recording_chunks, async () => {
    const file = await db.recording_files.get(id)
    // A chunk that arrives after recovery already finished or cancelled the capture
    // (or for a capture that never began) has nowhere left to go.
    if (!file || file.local_state !== 'capturing') return
    await db.recording_chunks.put({ recording_id: id, idx, blob })
    await db.recording_files.update(id, { last_chunk_at: Date.now() })
  })
}

/** The name a new recording gets: when it started, as the local `YYYY-MM-DD HH:MM`. */
export function defaultRecordingLabel(recordedAt: string | Date): string {
  const d = new Date(recordedAt)
  const pad = (n: number) => String(n).padStart(2, '0')
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ${pad(d.getHours())}:${pad(d.getMinutes())}`
}

export async function finishCapture(
  db: CrosstuneDb,
  id: string,
  fields: { tuneId: string | null; mime: string; durationMs: number; recordedAt: string },
): Promise<void> {
  await recordingTx(db, async () => {
    const file = await db.recording_files.get(id)
    // A repeat finish, or one after cancelCapture, has nothing left to assemble.
    if (!file || file.local_state !== 'capturing') return
    const chunks = await db.recording_chunks.where('recording_id').equals(id).sortBy('idx')
    const blob = new Blob(
      chunks.map((c) => c.blob),
      { type: fields.mime },
    )
    await db.recording_files.put(
      emptyFile(id, {
        blob,
        mime: fields.mime,
        bytes: blob.size,
        local_duration_ms: fields.durationMs,
      }),
    )
    await db.recording_chunks.where('recording_id').equals(id).delete()
    await putRecordingRow(db, id, {
      tuneId: fields.tuneId,
      source: 'microphone',
      // Named for when it started, so a recording is never nameless in a list.
      label: defaultRecordingLabel(fields.recordedAt),
      recordedAt: fields.recordedAt,
    })
  })
}

export async function cancelCapture(db: CrosstuneDb, id: string): Promise<void> {
  await recordingTx(db, async () => {
    const file = await db.recording_files.get(id)
    // A file that already finished (or was already cancelled) owns nothing here;
    // deleting it now would orphan the recordings row finishCapture just created.
    if (file && file.local_state !== 'capturing') return
    await db.recording_chunks.where('recording_id').equals(id).delete()
    if (file) await db.recording_files.delete(id)
  })
}

export async function addUploadedFile(
  db: CrosstuneDb,
  file: File,
  fields: { tuneId: string | null; label: string | null },
): Promise<string> {
  const id = newId()
  const mime = file.type || 'application/octet-stream'
  await recordingTx(db, async () => {
    await db.recording_files.put(emptyFile(id, { blob: file, mime, bytes: file.size }))
    await putRecordingRow(db, id, {
      tuneId: fields.tuneId,
      source: 'upload',
      label: fields.label,
      recordedAt: new Date(file.lastModified || Date.now()).toISOString(),
    })
  })
  return id
}

export async function updateRecording(
  db: CrosstuneDb,
  id: string,
  patch: { label?: string | null; tune_id?: string | null },
): Promise<void> {
  await recordingTx(db, async () => {
    const row = await db.recordings.get(id)
    if (!row || row.deleted_at) throw new Error(RECORDING_NOT_FOUND)
    let position = row.position
    if (patch.tune_id !== undefined && patch.tune_id !== row.tune_id && patch.tune_id) {
      const tune = await db.tunes.get(patch.tune_id)
      if (!tune || tune.deleted_at) throw new Error(TUNE_NOT_FOUND)
      position = nextPosition(await activeRecordingsForTune(db, patch.tune_id))
    }
    await putRow(db, 'recordings', {
      ...row,
      label: patch.label === undefined ? row.label : patch.label,
      tune_id: patch.tune_id === undefined ? row.tune_id : patch.tune_id,
      position,
      updated_at: now(),
    })
    // The edit re-pushes the row, so a failed upload gets a fresh attempt along with it.
    const file = await db.recording_files.get(id)
    if (file?.local_state === 'failed_upload') {
      await db.recording_files.update(id, { local_state: 'captured', error: null })
    }
  })
}

/** Put a refused or backed-off upload at the front of the queue, so the next pass tries it now. */
export async function retryUpload(db: CrosstuneDb, id: string): Promise<void> {
  await db.transaction('rw', db.recording_files, async () => {
    const file = await db.recording_files.get(id)
    if (file?.local_state !== 'failed_upload' && file?.local_state !== 'captured') return
    await db.recording_files.update(id, {
      local_state: 'captured',
      error: null,
      upload_attempts: 0,
      next_attempt_at: null,
    })
  })
}

export async function deleteRecording(db: CrosstuneDb, id: string): Promise<void> {
  await recordingTx(db, async () => {
    await tombstone(db, 'recordings', id, now())
    await db.recording_files.delete(id)
    await db.recording_chunks.where('recording_id').equals(id).delete()
  })
}

/** Called inside deleteTune's transaction, which is already widened to the audio tables. */
export async function tombstoneTuneRecordings(
  db: CrosstuneDb,
  tuneId: string,
  at: string,
): Promise<void> {
  const rows = await db.recordings.where('tune_id').equals(tuneId).toArray()
  for (const row of rows) {
    await tombstone(db, 'recordings', row.id, at, { enqueueDelete: false })
    await db.recording_files.delete(row.id)
    await db.recording_chunks.where('recording_id').equals(row.id).delete()
  }
}

export async function setFileState(
  db: CrosstuneDb,
  id: string,
  state: LocalFileState,
  error: string | null = null,
): Promise<void> {
  await db.recording_files.update(id, { local_state: state, error })
}

export async function storeDownloadedBlob(
  db: CrosstuneDb,
  id: string,
  blob: Blob,
  mime: string,
): Promise<void> {
  await db.transaction('rw', db.recording_files, async () => {
    const file = (await db.recording_files.get(id)) ?? emptyFile(id)
    await db.recording_files.put({
      ...file,
      blob,
      mime,
      bytes: blob.size,
      local_state: 'downloaded',
      error: null,
    })
  })
}

/** Drop cached audio that can be fetched again. Only a blob the server can
 * actually serve back (its recording is `ready`) exists anywhere else; a recording
 * still queued, blocked, failed, or not yet transcoded is the only copy. */
export async function clearDownloadedBlobs(db: CrosstuneDb): Promise<void> {
  await db.transaction('rw', db.recording_files, db.recordings, async () => {
    const candidates = await db.recording_files
      .filter((f) => f.local_state === 'uploaded' || f.local_state === 'downloaded')
      .toArray()
    const rows = await db.recordings.bulkGet(candidates.map((f) => f.id))
    for (const [i, file] of candidates.entries()) {
      const row = rows[i]
      if (!row || row.deleted_at || row.state !== 'ready') continue
      await db.recording_files.update(file.id, { blob: null, bytes: 0 })
    }
  })
}

export async function localAudioBytes(db: CrosstuneDb): Promise<number> {
  let bytes = 0
  // A cursor over the blob-bearing rows only, rather than materializing the whole table:
  // there is no index for "has a blob", so this still visits every row, but never holds
  // more than one in memory at a time.
  await db.recording_files
    .filter((f) => f.blob !== null)
    .each((f) => {
      bytes += f.bytes
    })
  return bytes
}

File: apple/CrosstuneKit/Sources/CrosstuneCommands/Recordings.swift
import CrosstuneStore
import Foundation
import GRDB

/// The name a new recording gets: when it started, as the local `YYYY-MM-DD HH:MM`. `timeZone`
/// defaults to the device's own so a caller only overrides it in a test.
public func defaultRecordingLabel(recordedAt: Timestamp, timeZone: TimeZone = .current) -> String {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone
    let parts = calendar.dateComponents([.year, .month, .day, .hour, .minute], from: recordedAt.date)
    return String(
        format: "%04d-%02d-%02d %02d:%02d", parts.year!, parts.month!, parts.day!, parts.hour!, parts.minute!)
}

/// A tune's recordings that have not been tombstoned, in position order.
private func activeRecordings(tuneID: String, db: Database) throws -> [Recording] {
    activeByPosition(try Recording.filter(Column("tune_id") == tuneID).fetchAll(db))
}

/// The type every finished capture is stored as.
public let capturedContentType = "audio/mp4"

extension StoreWriter {
    public func activeRecordingsForTune(_ tuneID: String) throws -> [Recording] {
        try activeRecordings(tuneID: tuneID, db: db)
    }

    /// Marks a capture as in progress before its first audio is written, so a capture the app
    /// never finished is found and recovered the next time the store opens.
    public func beginCapture(
        _ recordingID: String, fileName: String, tuneID: String?, recordedAt: Timestamp, at time: Timestamp = .now
    ) throws {
        try RecordingFile(
            id: recordingID, localState: .capturing, fileName: fileName, tuneID: tuneID, recordedAt: recordedAt,
            updatedAt: time
        ).save(db)
    }

    /// Turns a capture into a recording: its finished file becomes `captured`, waiting to upload,
    /// and the recording row is queued under the tune the capture began for, when that tune still
    /// exists. Returns false when the capture was already finished or cancelled.
    @discardableResult
    public func finishCapture(
        _ recordingID: String, fileName: String, bytes: Int64, durationMs: Int64, at time: Timestamp = .now
    ) throws -> Bool {
        guard let file = try RecordingFile.fetchOne(db, key: recordingID), file.localState == .capturing else {
            return false
        }
        let recordedAt = file.recordedAt ?? time
        let tuneID = try file.tuneID.flatMap { id in try Tune.fetchOne(db, key: id)?.deletedAt == nil ? id : nil }
        try RecordingFile(
            id: recordingID, localState: .captured, fileName: fileName, contentType: capturedContentType,
            bytes: bytes, localDurationMs: durationMs, updatedAt: time
        ).update(db)
        try putNewRecording(
            recordingID, tuneID: tuneID, source: "microphone",
            // Named for when it started, so a recording is never nameless in a list.
            label: defaultRecordingLabel(recordedAt: recordedAt), recordedAt: recordedAt, at: time)
        return true
    }

    /// Forgets a capture that is still in progress. A finished one is left alone, since its
    /// recording row already depends on it.
    public func cancelCapture(_ recordingID: String) throws {
        guard let file = try RecordingFile.fetchOne(db, key: recordingID), file.localState == .capturing else {
            return
        }
        try file.delete(db)
    }

    /// Adds a recording from an audio file already copied into the user's audio folder.
    public func addUploadedFile(
        _ recordingID: String = newID(), fileName: String, contentType: String?, bytes: Int64, tuneID: String?,
        label: String?, recordedAt: Timestamp, at time: Timestamp = .now
    ) throws -> String {
        try RecordingFile(
            id: recordingID, localState: .captured, fileName: fileName,
            contentType: contentType ?? "application/octet-stream", bytes: bytes, updatedAt: time
        ).save(db)
        try putNewRecording(
            recordingID, tuneID: tuneID, source: "upload", label: label, recordedAt: recordedAt, at: time)
        return recordingID
    }

    private func putNewRecording(
        _ recordingID: String, tuneID: String?, source: String, label: String?, recordedAt: Timestamp,
        at time: Timestamp
    ) throws {
        let siblings = try tuneID.map(activeRecordingsForTune) ?? []
        try put(
            Recording(
                id: recordingID, createdAt: time, tuneID: tuneID, source: source, recordedAt: recordedAt,
                label: label, position: nextPosition(siblings)),
            at: time)
    }

    /// Applies a label or tune change to a recording. Moving it to a different, live tune
    /// repositions it past that tune's other recordings; clearing its tune keeps its position.
    /// A failed upload is put back in the queue, since the edit re-pushes the row.
    public func updateRecording(
        _ recordingID: String, label: Patch<String?> = .keep, tuneID: Patch<String?> = .keep,
        at time: Timestamp = .now
    ) throws {
        guard var recording = try Recording.fetchOne(db, key: recordingID), recording.deletedAt == nil else {
            throw CommandError.recordingNotFound
        }
        if case .value(let newTuneID) = tuneID, let newTuneID, newTuneID != recording.tuneID {
            guard let tune = try Tune.fetchOne(db, key: newTuneID), tune.deletedAt == nil else {
                throw CommandError.tuneNotFound
            }
            recording.position = nextPosition(try activeRecordingsForTune(newTuneID))
        }
        recording.label = label.resolved(from: recording.label)
        recording.tuneID = tuneID.resolved(from: recording.tuneID)
        try put(recording, at: time)

        if var file = try RecordingFile.fetchOne(db, key: recordingID), file.localState == .failedUpload {
            file.localState = .captured
            file.error = nil
            file.updatedAt = time
            try file.update(db)
        }
    }

    /// Puts a file that failed to upload, or keeps failing, back at the front of the queue. The
    /// next transfer pass sends it.
    public func retryUpload(_ recordingID: String, at time: Timestamp = .now) throws {
        guard var file = try RecordingFile.fetchOne(db, key: recordingID),
            file.localState == .failedUpload || file.localState == .captured
        else { return }
        file.localState = .captured
        file.error = nil
        file.uploadAttempts = 0
        file.nextAttemptAt = nil
        file.updatedAt = time
        try file.update(db)
    }

    /// Lets go of audio this device can fetch again. Only a ready recording's audio is on the
    /// server to fetch; one still waiting, blocked, failed, or processing is the only copy.
    public func clearDownloadedAudio(at time: Timestamp = .now) throws {
        let candidates = try RecordingFile.filter(
            [LocalFileState.uploaded, .downloaded].contains(RecordingFile.CodingKeys.localState)
                && RecordingFile.CodingKeys.fileName != nil
        ).fetchAll(db)
        for var file in candidates {
            guard let row = try Recording.fetchOne(db, key: file.id), row.deletedAt == nil, row.state == "ready"
            else { continue }
            file.fileName = nil
            file.bytes = 0
            file.updatedAt = time
            try file.update(db)
        }
    }

    /// Tombstones a recording and drops its local audio file entry.
    public func deleteRecording(_ recordingID: String, at time: Timestamp = .now) throws {
        try tombstone(Recording.self, id: recordingID, at: time)
        try RecordingFile.deleteOne(db, key: recordingID)
    }
}

extension Commands {
    public func beginCapture(
        _ recordingID: String, fileName: String, tuneID: String?, recordedAt: Timestamp, at time: Timestamp = .now
    ) async throws {
        try await store.write { writer in
            try writer.beginCapture(recordingID, fileName: fileName, tuneID: tuneID, recordedAt: recordedAt, at: time)
        }
    }

    @discardableResult
    public func finishCapture(
        _ recordingID: String, fileName: String, bytes: Int64, durationMs: Int64, at time: Timestamp = .now
    ) async throws -> Bool {
        try await store.write { writer in
            try writer.finishCapture(recordingID, fileName: fileName, bytes: bytes, durationMs: durationMs, at: time)
        }
    }

    public func cancelCapture(_ recordingID: String) async throws {
        try await store.write { writer in try writer.cancelCapture(recordingID) }
    }

    @discardableResult
    public func addUploadedFile(
        _ recordingID: String = newID(), fileName: String, contentType: String?, bytes: Int64, tuneID: String?,
        label: String?, recordedAt: Timestamp, at time: Timestamp = .now
    ) async throws -> String {
        try await store.write { writer in
            try writer.addUploadedFile(
                recordingID, fileName: fileName, contentType: contentType, bytes: bytes, tuneID: tuneID, label: label,
                recordedAt: recordedAt, at: time)
        }
    }

    /// Recordings that are still capturing, as left by a capture the app never finished.
    public func capturingFiles() async throws -> [RecordingFile] {
        try await store.read { db in
            try RecordingFile.filter(RecordingFile.CodingKeys.localState == LocalFileState.capturing).fetchAll(db)
        }
    }

    public func activeRecordingsForTune(_ tuneID: String) async throws -> [Recording] {
        try await store.read { db in try activeRecordings(tuneID: tuneID, db: db) }
    }

    public func updateRecording(
        _ recordingID: String, label: Patch<String?> = .keep, tuneID: Patch<String?> = .keep,
        at time: Timestamp = .now
    ) async throws {
        try await store.write { writer in
            try writer.updateRecording(recordingID, label: label, tuneID: tuneID, at: time)
        }
    }

    public func retryUpload(_ recordingID: String, at time: Timestamp = .now) async throws {
        try await store.write { writer in try writer.retryUpload(recordingID, at: time) }
    }

    /// Deletes the audio of every ready recording from this device. It downloads again when
    /// played.
    public func clearDownloadedAudio(at time: Timestamp = .now) async throws {
        try await store.writeDroppingAudio { writer in try writer.clearDownloadedAudio(at: time) }
    }

    /// Deletes a recording, its local file row, and its audio on this device.
    public func deleteRecording(_ recordingID: String, at time: Timestamp = .now) async throws {
        try await store.writeDroppingAudio { writer in try writer.deleteRecording(recordingID, at: time) }
    }
}
